packages/flutter/lib/src/cupertino/text_selection_toolbar.dart

```diff
@@ -1021,15 +1059,24 @@ class _RenderCupertinoTextSelectionToolbarItems extends RenderBox with Container
       }
     });
 
+    // Re-layout the children using max. child height to allow them to be centered vertically.
+    visitChildren((RenderObject renderObjectChild) {
```

Comment:
We should avoid doing a second layout pass if possible. Do these children all have to have the exact same height? To center them vertically could you just adjust their y offsets instead of doing a second layout pass and enforcing `greatestHeight`?

Reply:
Also unless `visitChildren` is overridden it doesn't visit the back button and the next button right? Do they need to readjust their heights too?

Reply:
Yes, we need to adjust the children height so the whole button area is painted using another shade during tap - see the screenshot [here](https://github.com/flutter/flutter/pull/133386#discussion_r1331144856) (the bottom toolbar). `visitChildren` also visits the back/next buttons: they're added as render object children (and called "slotted" children) and their heigh is adjusted.

Reply:
Ah ok that's unfortunate. Instead of  `RenderObject.layout`, usually `getMaxIntrinsicHeight` is used to do this kind of speculative layout. 

the back/next buttons seem to be added to the parent via `adoptChild`. This doesn't add them to the linked list of children that `visitChildren` traverses.

Reply:
The `visitChildren` method is overriden though:
```dart
  @override
  void visitChildren(RenderObjectVisitor visitor) {
    // Visit the slotted children.
    if (_backButton != null) {
      visitor(_backButton!);
    }
    if (_nextButton != null) {
      visitor(_nextButton!);
    }
    // Visit the list children.
    super.visitChildren(visitor);
  }
```